File: tests/r_parity/26_glmm_logit.py
"""StatsPAI GLMM logit parity (Python side) -- Module 26.

Random-intercept logistic GLMM on a deterministic n=1500, 60-group
panel. The companion 26_glmm_logit.R uses lme4::glmer with family =
binomial. Tolerance: rel < 2e-4 on fixed-effect point estimates and
rel < 5e-2 on SE because Laplace fixed-effect covariance conventions
and R/Stata likelihood implementations differ slightly; AGHQ is covered
in Module 27.
"""
from __future__ import annotations

import numpy as np
import pandas as pd
import statspai as sp

from _common import PARITY_SEED, ParityRecord, dump_csv, write_results


MODULE = "26_glmm_logit"


def make_data(n_groups: int = 60, n_per: int = 25,
              seed: int = PARITY_SEED) -> pd.DataFrame:
    rng = np.random.default_rng(seed)
    gid = np.repeat(np.arange(n_groups), n_per)
    n = n_groups * n_per
    x1 = rng.normal(size=n)
    group_re = rng.normal(scale=0.8, size=n_groups)
    linpred = -0.5 + 0.7 * x1 + group_re[gid]
    prob = 1 / (1 + np.exp(-linpred))
    y = (rng.uniform(size=n) < prob).astype(int)
    return pd.DataFrame({"y": y, "x1": x1, "gid": gid})


def main() -> None:
    df = make_data()
    dump_csv(df, MODULE)

    fit = sp.melogit(data=df, y="y", x_fixed=["x1"], group="gid")

    rows: list[ParityRecord] = [
        ParityRecord(MODULE, "py", "beta_intercept",
                     estimate=float(fit.params["_cons"]),
                     se=float(fit.bse["_cons"]),
                     n=int(fit.n_obs)),
        ParityRecord(MODULE, "py", "beta_x1",
                     estimate=float(fit.params["x1"]),
                     se=float(fit.bse["x1"]),
                     n=int(fit.n_obs)),
        ParityRecord(MODULE, "py", "logLik",
                     estimate=float(fit.log_likelihood),
                     n=int(fit.n_obs)),
    ]

    write_results(MODULE, "py", rows,
                  extra={"family": "binomial",
                         "link": "logit",
                         "n_groups": int(fit.n_groups),
                         "nAGQ": 1,
                         "optimizer_tol": 1e-8,
                         "optimizer_note": (
                             "sp.melogit uses the tightened default "
                             "tol=1e-8 so the Laplace likelihood optimum "
                             "tracks lme4/Stata fixed effects."
                         )})


if __name__ == "__main__":
    main()

File: tests/r_parity/26_glmm_logit.R
# StatsPAI GLMM logit parity (R side) -- Module 26.
#
# Reads data/26_glmm_logit.csv and runs lme4::glmer with family =
# binomial(link = "logit"), Laplace approximation (nAGQ = 1).
# Tolerance: rel < 2e-4 on fixed-effect point estimates and rel < 5e-2
# on SE because Laplace fixed-effect covariance conventions and
# R/Stata likelihood implementations differ slightly.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(lme4)
})

MODULE <- "26_glmm_logit"

df <- read_csv_strict(MODULE)
df$gid <- as.factor(df$gid)

fit <- lme4::glmer(y ~ x1 + (1 | gid), data = df,
                    family = binomial(link = "logit"),
                    nAGQ = 1L)

co  <- fixef(fit)
ses <- sqrt(diag(vcov(fit)))
ll  <- as.numeric(logLik(fit))

rows <- list(
  parity_row(MODULE, "beta_intercept",
             estimate = unname(co["(Intercept)"]),
             se = unname(ses["(Intercept)"]),
             n = nrow(df)),
  parity_row(MODULE, "beta_x1",
             estimate = unname(co["x1"]),
             se = unname(ses["x1"]),
             n = nrow(df)),
  parity_row(MODULE, "logLik", estimate = ll, n = nrow(df))
)

write_results(MODULE, rows,
              extra = list(family = "binomial(logit)",
                           nAGQ = 1L,
                           n_groups = length(unique(df$gid))))
